### PR TITLE
[FIX] account_edi - revert to using AccountTax._compute_all to compute tax amounts corresponding to each invoice line in the edi document for the invoice.

### PR DESCRIPTION
**Pull request by antoine162**
#### The issue

In Mexico, companies are required to send every invoice in electronic format to the government. This document (called the CFDI) specifies, for each invoice line, the taxes applicable, the amount before tax and the tax amount. The document also specifies the total amounts withheld for each tax which was applicable to the invoice. 

The PAC (the external service which signs this electronic document and sends it to the government) does checks on those amounts. In particular, the tax amounts must satisfy two constraints to pass validation:

(1) The total tax amount for a tax must be equal to the sum of the tax amounts reported for each invoice line.
(2) The tax amount reported for each line must be equal to (tax rate * base amount), rounded either up or down. For example, for a line with base = MXN 398.28 and an applicable VAT of 16%, the exact tax amount would be 0.16 * 398.28 = 63.7248, so the acceptable values for the tax amount on that line are 63.72 and 63.73.

In v14, we used the AccountTax.compute_all method to compute the tax amounts reported for each invoice line. This gave the exact tax amount, rounded (either up or down depending on which side of 0.5 it is). This always fulfils condition (2) but fulfils condition (1) only if the tax rounding method is 'Round Locally'. Because of this, in v14 MX users couldn't use 'Round Globally'.

In v15, we use a new SQL-based method to compute the tax amount reported for each invoice line: we take the total tax amount, and allocate it among the invoice lines proportionately to the base amount of the invoice line. This always fulfils condition (1) but sometimes doesn't fulfil condition (2). This has caused many issues in customer DBs (see list of tickets at the end). 

This was introduced in 433656415a3c and seems to be an issue for all MX customers who have upgraded to v15. 

#### The solution
@smetl and I have looked into what can be done about this. 

- Modifying the tax amount computation method in a way which satisfies both conditions is possible, but requires more work and testing. 
- The existing v14 code generates tax amounts which satisfy both conditions, at least when the tax computation method is 'Round Locally'. 
- A good temporary solution, while we try to modify the tax amount computation to satisfy both conditions, would therefore be to use the v14 tax amount computation when the tax rounding method is 'Round Locally'. While not ideal, this will at least enable customers with the 'Round Locally' rounding method to correctly submit their invoices to the government. Customers with the 'Round Globally' method will need to wait for the full fix. 

This PR does exactly that. 

#### Tickets linked to this problem (non-exhaustive list):
opw-2695243
opw-2733280
opw-2723571
opw-2724805
opw-2722370
opw-2722388

### CODE REVIEW

**Comment by antoine162** [comment]
@smetl

File: addons/account_edi/models/account_move.py [hunk 1]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
...
+                        'tax_amount_currency': sign * tax_res['amount'],
+                    })
+            return invoice_lines_tax_values_dict
+
+        # Decide whether we will use AccountTax.compute_all or AccountMoveLine._get_query_tax_details to compute
+        # the tax amounts for each invoice.
+        # By default: if using 'Round Locally', we use compute_all (which will give rounded amounts for each line)
+        #             if using 'Round Globally', we use _get_query_tax_details (which will allocate the total tax amount proportionately among the lines)
+        # This can be changed using the 'use_sql_query_tax_details' context key.
+        company = self[0].company_id if self else self.env.company_id
```
Comment (FlorianGilbert) [inline addons/account_edi/models/account_move.py:287]: You make a `self.ensure_one()` [in line 221](https://github.com/odoo/odoo/pull/84312/files#diff-50d4c0850ad46ccdda325a6de23e3232b5a1e42cfb04238b08188c0bf50132b4R221). So the `self[0].company_id if self` is useless no ?
```suggestion
        company = self.company_id or self.env.company_id
```

File: addons/account_edi/models/account_move.py [hunk 2]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
...
+                src_line = invoice_lines.browse(row['src_line_id'])
+
+                tax = self.env['account.tax'].browse(row['tax_id'])
+
+                invoice_lines_tax_values_dict[invoice_line].append({
+                    'base_line_id': invoice_line,
+                    'tax_line_id': tax_line,
+                    'src_line_id': src_line,
+                    'tax_id': tax,
+                    'src_tax_id': self.env['account.tax'].browse(row['group_tax_id']) if row['group_tax_id'] else tax,
```
Comment (FlorianGilbert) [inline addons/account_edi/models/account_move.py:246]: Why don't you do the same as for `invoice_line`, `tax_line`, `src_line` and `tax` ?

File: addons/account_edi/models/account_move.py [hunk 3]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
...
+        def compute_invoice_lines_tax_values_dict_from_sql_query(invoice_lines):
+            invoice_lines_tax_values_dict = defaultdict(list)
+            domain = [('move_id', '=', self.id)]
+            tax_details_query, tax_details_params = invoice_lines._get_query_tax_details_from_domain(domain)
+            self._cr.execute(tax_details_query, tax_details_params)
+            for row in self._cr.dictfetchall():
+                invoice_line = invoice_lines.browse(row['base_line_id'])
+                tax_line = invoice_lines.browse(row['tax_line_id'])
+                src_line = invoice_lines.browse(row['src_line_id'])
+
```
Comment (FlorianGilbert) [inline addons/account_edi/models/account_move.py:238]: You can remove this line.

**Review by FlorianGilbert — CHANGES_REQUESTED** [review]

File: addons/account_edi/models/account_move.py [hunk 4]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
...
+                    'src_line_id': src_line,
+                    'tax_id': tax,
+                    'src_tax_id': self.env['account.tax'].browse(row['group_tax_id']) if row['group_tax_id'] else tax,
+                    'tax_repartition_line_id': tax_line.tax_repartition_line_id,
+                    'base_amount': row['base_amount'],
+                    'tax_amount': row['tax_amount'],
+                    'base_amount_currency': row['base_amount_currency'],
+                    'tax_amount_currency': row['tax_amount_currency'],
+                })
+            invoice_lines_tax_values_dict.default_factory = None
```
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account_edi/models/account_move.py:253]: That's true. (I seemed to remember that it was good practice when exposing a defaultdict, but here I don't think it's needed.)

File: addons/account_edi/models/account_move.py [hunk 5]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
         def default_grouping_key_generator(tax_values):
             return {'tax': tax_values['tax_id']}
 
+        def compute_invoice_lines_tax_values_dict_from_sql_query(invoice_lines):
+            invoice_lines_tax_values_dict = defaultdict(list)
+            domain = [('move_id', '=', self.id)]
+            tax_details_query, tax_details_params = invoice_lines._get_query_tax_details_from_domain(domain)
```
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account_edi/models/account_move.py:232]: :+1:

File: addons/account_edi/models/account_move.py [hunk 6]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
...
+                    })
+            return invoice_lines_tax_values_dict
+
+        # Decide whether we will use AccountTax.compute_all or AccountMoveLine._get_query_tax_details to compute
+        # the tax amounts for each invoice.
+        # By default: if using 'Round Locally', we use compute_all (which will give rounded amounts for each line)
+        #             if using 'Round Globally', we use _get_query_tax_details (which will allocate the total tax amount proportionately among the lines)
+        # This can be changed using the 'use_sql_query_tax_details' context key.
+        company = self[0].company_id if self else self.env.company_id
+        use_sql_query_tax_details = True if company.tax_calculation_rounding_method == 'round_globally' else False
```
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account_edi/models/account_move.py:288]: :+1:

File: addons/account_edi/models/account_move.py [hunk 7]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
...
-                'src_tax_id': self.env['account.tax'].browse(row['group_tax_id']) if row['group_tax_id'] else tax,
-                'tax_repartition_line_id': tax_line.tax_repartition_line_id,
-                'base_amount': row['base_amount'],
-                'tax_amount': row['tax_amount'],
-                'base_amount_currency': row['base_amount_currency'],
-                'tax_amount_currency': row['tax_amount_currency'],
-            })
+        invoice_lines_tax_values_dict = compute_invoice_lines_tax_values_dict_from_sql_query(invoice_lines) \
+                                        if use_sql_query_tax_details \
+                                        else compute_invoice_lines_tax_values_dict_from_compute_all(invoice_lines)
```
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account_edi/models/account_move.py:299]: :+1:

File: addons/account_edi/models/account_move.py [hunk 8]
```diff
@@ -359,8 +402,10 @@ def default_grouping_key_generator(tax_values):
                 else:
                     invoice_line_global_tax_details = invoice_global_tax_details['invoice_line_tax_details'][invoice_line]
 
-                self._add_edi_tax_values(invoice_global_tax_details, grouping_key, serialized_grouping_key, tax_values, key_by_tax)
-                self._add_edi_tax_values(invoice_line_global_tax_details, grouping_key, serialized_grouping_key, tax_values, key_by_tax)
+                self._add_edi_tax_values(invoice_global_tax_details, grouping_key, serialized_grouping_key, tax_values,
+                                         key_by_tax=key_by_tax if use_sql_query_tax_details else None)
+                self._add_edi_tax_values(invoice_line_global_tax_details, grouping_key, serialized_grouping_key, tax_values,
+                                         key_by_tax=key_by_tax if use_sql_query_tax_details else None)
```
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account_edi/models/account_move.py:408]: I think the only issue with this is that at line 393 we need `key_by_tax` to be a dict:

```py
key_by_tax[tax_values['tax_id']] = serialized_grouping_key
```

On balance, I think the logic is clearer if we leave things as is - i.e. we construct `key_by_tax`, but pass it to `_add_edi_tax_values` only if `use_sql_query_tax_details` is True.

What do you think?

File: addons/account_edi/models/account_move.py [hunk 9]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
...
+                for tax_res in taxes_res['taxes']:
+                    tax_values_list.append({
+                        'base_line_id': invoice_line,
+                        'tax_id': self.env['account.tax'].browse(tax_res['id']),
+                        'tax_repartition_line_id': self.env['account.tax.repartition.line'].browse(tax_res['tax_repartition_line_id']),
+                        'base_amount': sign * invoice_line.company_currency_id.round(tax_res['base'] * rate),
+                        'tax_amount': sign * invoice_line.company_currency_id.round(tax_res['amount'] * rate),
+                        'base_amount_currency': sign * tax_res['base'],
+                        'tax_amount_currency': sign * tax_res['amount'],
+                    })
```
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account_edi/models/account_move.py:279]: That is absolutely true!!! Thanks - I've changed it.

File: addons/account_edi/models/account_move.py (hunk 1 above)
Comment (antoine162) [reply addons/account_edi/models/account_move.py:287]: Heh, yes that's right. And thank you as well for spotting the case where `self.company_id` is not set.

File: addons/account_edi/models/account_move.py (hunk 2 above)
Comment (antoine162) [reply addons/account_edi/models/account_move.py:246]: That makes sense - I don't know why it was this way (I just reused the existing code), but I've changed it now.

File: addons/account_edi/models/account_move.py (hunk 3 above)
Comment (antoine162) [reply addons/account_edi/models/account_move.py:238]: :+1:

File: addons/account_edi/models/account_move.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (FlorianGilbert) [reply addons/account_edi/models/account_move.py:407]: Oh my bad! 
The other solution is to use a temporary variable but as you want :wink:

File: addons/account_edi/models/account_move.py [hunk 10]
```diff
@@ -226,35 +226,78 @@ def _serialize_python_dictionary(vals):
...
+            return invoice_lines_tax_values_dict
+
+        # Decide whether we will use AccountTax.compute_all or AccountMoveLine._get_query_tax_details to compute
+        # the tax amounts for each invoice.
+        # By default: if using 'Round Locally', we use compute_all (which will give rounded amounts for each line)
+        #             if using 'Round Globally', we use _get_query_tax_details (which will allocate the total tax amount proportionately among the lines)
+        # This can be changed using the 'use_sql_query_tax_details' context key.
+        company = self[0].company_id if self else self.env.company_id
+        use_sql_query_tax_details = True if company.tax_calculation_rounding_method == 'round_globally' else False
+        if 'use_sql_query_tax_details' in self.env.context:
```
[the remark this answers is not in the document]
Comment (FlorianGilbert) [reply addons/account_edi/models/account_move.py:287]: It's better! You could also doing something like this
```python
use_sql_query_tax_details = bool(self.env.context.get('use_sql_query_tax_details', False))
```
As you want :wink:

File: addons/account_edi/models/account_move.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (antoine162) [reply addons/account_edi/models/account_move.py:407]: My gut feeling is that it's slightly clearer without the temporary variable. Happy to do either.

**Review by FlorianGilbert — APPROVED** [review]
@robodoo r+

**Comment by FlorianGilbert** [comment]
@robodoo retry

**Comment by FlorianGilbert** [comment]
@antoine162 :point_up: Could you check it and fix it please ?
May be a rebase is sufficient

**Comment by antoine162** [comment]
Coucou @FlorianGilbert it was actually a bug. Basically:

before this PR, in the `add_edi_tax_values` function, there was this code:
```py
if key_by_tax[tax_values['tax_id']] != key_by_tax.get(tax_values['src_line_id'].tax_line_id):
    tax_details['base_amount'] += tax_values['base_amount']
    tax_details['base_amount_currency'] += tax_values['base_amount_currency']
```

The reason for this code is that when using the SQL query to retrieve the tax amounts, if a first tax affects the base for a second tax, the SQL query will return two separate rows for the second tax.

E.g. suppose the account.move is as follows

```
   line    |       tax_ids       |    tax_line_id   |   balance   |
-------------------------------------------------------------------
base_line  |  10_affect_base, 20 |                  |      200    |
tax_line_1 |                     |  10_affect_base  |       20    |
tax_line_2 |                     |        20        |       44    |
```

Then the SQL query returns the following rows:
```
 base_line |  src_line  |  tax_line   |     tax_id     |   tax_amount   |
-------------------------------------------------------------------------
base_line  |  base_line |  tax_line_1 | 10_affect_base |             20 |
base_line  |  base_line |  tax_line_2 |        20      |             40 |
base_line  | tax_line_1 |  tax_line_1 |        20      |              4 |
```

Because of this, the `key_by_tax` dictionary prevents the base amount from being added twice when there are several rows corresponding to the same base line (but some of these rows are due to taxes which affected the base amount). 

In contrast, when we use the _compute_all method on the account.move, it returns the following values:
```
 base_line |  tax_line   |     tax_id     |   tax_amount   |
------------------------------------------------------------
base_line  |  tax_line_1 | 10_affect_base |             20 |
base_line  |  tax_line_2 |       20       |             44 |
```

So with the _compute_all method, there isn't any need to prevent the base_amount from being summed up (and therefore no need for the `key_by_tax` dict. 

I've therefore changed the condition to
```py
if not (key_by_tax and key_by_tax[tax_values['tax_id']] == key_by_tax.get(tax_values['src_line_id'].tax_line_id)):
    tax_details['base_amount'] += tax_values['base_amount']
    tax_details['base_amount_currency'] += tax_values['base_amount_currency']
```
I hope you can see from the explanation above why this fix is correct (but don't hesitate to ask if it's unclear). 

I've verified that with this fix the failing UT (`odoo/addons/l10n_es_edi_sii/tests/test_edi_xml.py`) now passes.

**Comment by antoine162** [comment]
@robodoo retry

File: addons/account_edi/models/account_move.py [hunk 11]
```diff
@@ -226,35 +226,75 @@ def _serialize_python_dictionary(vals):
...
+                    })
+            return invoice_lines_tax_values_dict
+
+        # Decide whether we will use AccountTax.compute_all or AccountMoveLine._get_query_tax_details to compute
+        # the tax amounts for each invoice.
+        # By default: if using 'Round Locally', we use compute_all (which will give rounded amounts for each line)
+        #             if using 'Round Globally', we use _get_query_tax_details (which will allocate the total tax amount proportionately among the lines)
+        # This can be changed using the 'use_sql_query_tax_details' context key.
+        company = self.company_id or self.env.company_id
+        use_sql_query_tax_details = bool(self.env.context.get('use_sql_query_tax_details', company.tax_calculation_rounding_method == 'round_globally'))
```
Comment (smetl) [inline addons/account_edi/models/account_move.py:286]: add an explicit parameter. This issue is not related to the `round_globally` method

File: addons/account_edi/models/account_move.py [hunk 12]
```diff
@@ -226,35 +226,75 @@ def _serialize_python_dictionary(vals):
...
+                        'tax_amount_currency': sign * tax_res['amount'],
+                    })
+            return invoice_lines_tax_values_dict
+
+        # Decide whether we will use AccountTax.compute_all or AccountMoveLine._get_query_tax_details to compute
+        # the tax amounts for each invoice.
+        # By default: if using 'Round Locally', we use compute_all (which will give rounded amounts for each line)
+        #             if using 'Round Globally', we use _get_query_tax_details (which will allocate the total tax amount proportionately among the lines)
+        # This can be changed using the 'use_sql_query_tax_details' context key.
+        company = self.company_id or self.env.company_id
```
Comment (smetl) [inline addons/account_edi/models/account_move.py:285]: self.company_id is always set. Then, your `or` is useless

**Comment by smetl** [comment]
Also, plz add a test for mx showcasing your issue.

**Comment by smetl** [comment]
@robodoo r-

**Comment by antoine162** [comment]
> Also, plz add a test for mx showcasing your issue.

Here's the l10n_mx_edi UT: https://github.com/odoo/enterprise/pull/24261

File: addons/account_edi/models/account_move.py (hunk 11 above)
Comment (antoine162) [reply addons/account_edi/models/account_move.py:287]: @smetl actually, the new 15.0 method causes issues only when the rounding method is set to 'Round per Line'. AFAIK it works fine with 'Round Globally'. 

Conversely, the 14.0 method didn't handle well 'Round Globally'.

(see https://www.odoo.com/web#id=2593016&cids=1&menu_id=4720&action=333&active_id=49&model=project.task&view_type=form)

So doesn't it makes sense to keep the 15.0 method when the user uses 'Round Globally', and use the 14.0 method when the user uses 'Round Locally'?

File: addons/account_edi/models/account_move.py (hunk 12 above)
Comment (antoine162) [reply addons/account_edi/models/account_move.py:286]: Thanks! Now I know that company_id is always set on an account.move. I've applied your change in the code.

File: addons/account_edi/models/account_move.py (hunk 11 above)
Comment (antoine162) [reply addons/account_edi/models/account_move.py]: Happy to change the context key into an explicit parameter though if you feel that's better :smile:

**Comment by antoine162** [comment]
The l10n_mx_edi UT -> https://github.com/odoo/enterprise/pull/24656

File: addons/account_edi/models/account_move.py (hunk 11 above)
Comment (antoine162) [reply addons/account_edi/models/account_move.py]: Done :+1:

**Comment by smetl** [comment]
@robodoo r+
